### PR TITLE
feat: `extensions_options` macro

### PR DESCRIPTION
**Pull request by crepererum**
# Which issue does this PR close?
\-

# Rationale for this change
I was implementing `ExtensionsOptions` in InfluxDB IOx and it was quite cumbersome, since you need to be very careful to keep all the `set`/`entries` methods in sync with the structure definition. So I wrote a macro that does a similar magic than the DF builtin configs. I would like to upstream this since I think it actually makes config extensions usable.

# What changes are included in this PR?
A new macro.

# Are these changes tested?
Small "does it expand" test.

# Are there any user-facing changes?
New macro.

### CODE REVIEW

**Review by waynexia — APPROVED** [review]
Looks good to me!

The default value might be better with an attribute macro, but it would be much more complicated. This one looks pretty enough :+1:

**Comment by crepererum** [comment]
The issue with prog macros (which I think is what you mean with "attribute macros") is: they are slower, more complicated to write & test, and you ideally have a trait just for the prog macro because they are compiled just to check/document the downstream and you wanna avoid that for costly crates.

If you meant "can you just use a macro syntax for defaults"? Ahh, maybe. If there's strong desire to do that, I can try (although I'm not sure if traditional macros can match on macro-like attributes).

**Comment by waynexia** [comment]
>The issue with prog macros (which I think is what you mean with "attribute macros") is: they are slower, more complicated to write & test, and you ideally have a trait just for the prog macro because they are compiled just to check/document the downstream and you wanna avoid that for costly crates.

Agree, I tried one recently and it does take me lots of time to debug :cry: 

>If you meant "can you just use a macro syntax for defaults"? Ahh, maybe. If there's strong desire to do that, I can try (although I'm not sure if traditional macros can match on macro-like attributes).

As long as the current implementation works fine and does not have ambiguous grammar, we can keep using it. And to me it's straitforward :+1:

File: datafusion/core/tests/config.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::common::extensions_options;
+
+// The structure is not really used. We just check that the macro re-export works and everything referenced within the
```
Comment (alamb) [inline datafusion/core/tests/config.rs:20]: Can you please move this code to https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples/examples

That would also provide a good place maybe to provide some simple program showing how to use options created with this macro 



Maybe we could add a small test showing that the configuration could be instantiated and used 🤔  As a way to try and document its use.

**Review by alamb — COMMENTED** [review]
Other than the location of the test I think this one is ready to go -- thank you @crepererum 

For anyone following along, the config was added to IOx in https://github.com/influxdata/influxdb_iox/pull/7020/files

File: datafusion/core/tests/config.rs (hunk 1 above)
Comment (crepererum) [reply datafusion/core/tests/config.rs]: done

File: datafusion-examples/examples/config_extension.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+impl ConfigExtension for MyConfig {
+    const PREFIX: &'static str = "my_config";
+}
+
+fn main() {
+    // set up config struct and register extension
+    let mut config = ConfigOptions::default();
+    config.extensions.insert(MyConfig::default());
```
Comment (alamb) [inline datafusion-examples/examples/config_extension.rs:43]: ❤️

File: datafusion/common/src/config.rs [hunk 3]
```diff
@@ -634,3 +634,129 @@ trait Visit {
 
     fn none(&mut self, key: &str, description: &'static str);
 }
+
+/// Convenience macro to create [`ExtensionsOptions`].
```
Comment (alamb) [inline datafusion/common/src/config.rs:638]: giphy

**Review by alamb — APPROVED** [review]
Thank you @crepererum  -- this is really nice. Thank you @waynexia  for the review

**Pull request merged**